crates/syn2mas/src/synapse_reader/checks.rs

```diff
@@ -0,0 +1,290 @@
+    NonEmailThreepidsInDatabase { num_non_email_3pids: i64 },
+
+    #[error(
+        "Synapse config has `enable_3pid_changes` explicitly enabled, which must be disabled or removed."
+    )]
+    ThreepidChangesEnabled,
+
+    #[error("Synapse config has `login_via_existing_session.enabled` set to true, which must be disabled.")]
+    LoginViaExistingSessionEnabled,
+
+    #[error("MAS configuration has the wrong `matrix.homeserver` set ({mas:?}), it should match Synapse's `server_name` ({synapse:?})")]
+    ServerNameMismatch { synapse: String, mas: String },
+
+    #[error("Synapse database contains {num_users} users associated to the OpenID Connect or OAuth2 provider '{provider}' but the Synapse configuration does not contain this provider.")]
+    SynapseMissingOAuthProvider { provider: String, num_users: i64 },
+
+    #[error("Synapse config contains an OpenID Connect or OAuth2 provider '{provider}' (issuer: {issuer:?}) used by {num_users} users which must also be configured in the MAS configuration as an upstream provider.")]
+    MasMissingOAuthProvider {
+        provider: String,
+        issuer: String,
+        num_users: i64,
+    },
+}
+
+/// A potential hazard found whilst checking the Synapse database, that should be presented
+/// to the operator to check they are aware of a caveat before proceeding with the migration.
+#[derive(Debug, Error)]
+pub enum CheckWarning {
+    #[error("Synapse config contains OIDC auth configuration (issuer: {issuer:?}) which will need to be manually mapped to an upstream OpenID Connect Provider during migration.")]
+    UpstreamOidcProvider { issuer: String },
+
+    #[error("Synapse config contains {0} auth configuration which will need to be manually mapped as an upstream OpenID Connect Provider during migration.")]
+    ExternalAuthSystem(&'static str),
+
+    #[error("Synapse config has registration enabled. This must be disabled after migration before bringing Synapse back online.")]
+    DisableRegistrationAfterMigration,
+
+    #[error("Synapse config has `user_consent` enabled. This should be disabled after migration.")]
+    DisableUserConsentAfterMigration,
+
+    #[error("Synapse config has a registration CAPTCHA enabled. An equivalent CAPTCHA should be manually configured in MAS.")]
+    ShouldPortRegistrationCaptcha,
+}
+
+/// Check that the Synapse configuration is sane for migration.
+#[must_use]
+pub fn synapse_config_check(synapse_config: &Config) -> (Vec<CheckWarning>, Vec<CheckError>) {
+    let mut errors = Vec::new();
+    let mut warnings = Vec::new();
+
+    if synapse_config.enable_registration {
+        warnings.push(CheckWarning::DisableRegistrationAfterMigration);
+    }
+    if synapse_config.enable_registration_captcha {
+        warnings.push(CheckWarning::ShouldPortRegistrationCaptcha);
+    }
+    if synapse_config.user_consent {
+        warnings.push(CheckWarning::DisableUserConsentAfterMigration);
+    }
```

Comment:
I wonder if for those two we should look at the MAS config? Checking whether CAPTCHA is enabled is straightforward. User consent, the only equivalent we have in MAS is the `terms_url`

Reply:
1c6914616bbce812f9c61c5379840376c27e7bb5

For user consent we still emit a warning when enabled since I guess we don't want Synapse to lock out the user until it's been agreed on the Synapse side.